platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,852 @@
+  - Manual save
+  - Auto save
+  - Data migrations
+
+**Each sub-module will be independent.** If a form doesn't need the multi-page
+routing, the team building it simply won't use the routing sub-module, and it
+won't be included in their application bundle. Likewise, if a form needs more
+complex routing, the team building it can opt to roll their own, but still reap
+the benefits of using the rest of the system.
+
+_Note: "sub-modules" here refer to independent parts of a larger system and have
+nothing to do with [Git
+Sub-Modules](https://git-scm.com/book/en/v2/Git-Tools-Sub-Modules)._
+
+Additionally, the library will provide a number of helper functions and
+components for:
+- Field-level schema validation
+- [Building a review page](#json-schema-validation)
+
+## Specifics
+
+### Detailed Design
+**Key technologies** we'll use:
+- React
+- TypeScript
+- [`api-extractor`](https://api-extractor.com/pages/overview/intro/)
+- [React router](https://reactrouter.com/)
+- [Formik](https://formik.org/)
+
+The following sections outline the various parts of the system.
+
+#### Form page builder
+Formulate will leverage Formik for much of the form page building. It will
+**re-export Formik components** where it makes sense and **provide light
+wrappers for Formik input fields** using design system components.
+
+In addition to simple wrappers, Formulate will use Formik to provide more
+complex fields such as file upload which may be imported and used in
+applications.
+
+##### Why not just use Formik directly?
+It should be possible to use Formik directly without any issues. The form page
+builder sub-module intends to make using the design system with Formik simpler.
+
+#### Routing
+This sub-module will be responsible for the user navigating through a multi-page
+form. In particular, it will:
+- Manage routes for each page
+- Render each page wrapped in the chrome appropriate for navigation
+  - Form title, page title, progress bar, back / continue buttons, etc.
+- Prevent progress through the form if there are validation errors on the
+  current page
+- Manage the form data
+  - Each individual page is essentially its own mini-form; the router stitches
+    all the data together into a single cohesive experience
+
+To accomplish this, there will be four primary pieces in the routing
+sub-module:
+- [`Router` component](#router-component)
+- [`Chapter` component](#chapter-component)
+- [`Page` component](#page-component)
+- [`Route` component](#route-component)
+
+Additionally, to accommodate some common patterns on VA.gov form applications,
+Formulate will provide:
+- [Helpers for building array pages](#functionality-array-pages)
+- [Helpers for buliding the review page](#review-page)
+
+##### `Router` Component
+**Purpose:**
+The `Router` is at the heart of the routing sub-module. It will hold the routes
+and components to render at each route.
+
+**Usage requirements:**
+_None_
+
+**Responsibilities:**
```

Comment:
I guess `Purpose` is more like `Description`. It's meant to be a brief description of what the component does. Its `Responsibilities` are a list of things it will do.

I'm open to changing up the format, but the idea here was to split out answering the "What is it?" question and the "How does it work?" question. I decided not to go with `What is it?` and `How does it work?` because I didn't want to get _too_ low-level. I felt like it was going to be wordy enough as it was. :sweat_smile: :grimacing:

Reply:
I think I understand the difference now with a bit more noodling.